pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,1100 @@
+		} else {
+			latestToken, err := tokenInterface.Get(userToken.Name, metav1.GetOptions{})
+			if err != nil {
+				logrus.Errorf("[%v] token %s no longer exists: %v", migrateTokensOperation, userToken.Name, err)
+			}
+			if latestToken.Annotations == nil {
+				latestToken.Annotations = make(map[string]string)
+			}
+			latestToken.Annotations[adGUIDMigrationAnnotation] = workunit.guid
+			if latestToken.Labels == nil {
+				latestToken.Labels = make(map[string]string)
+			}
+			latestToken.Labels[tokens.UserIDLabel] = workunit.originalUser.Name
+			latestToken.Labels[adGUIDMigrationLabel] = migratedLabelValue
+			latestToken.UserPrincipal.Name = localPrincipalID
+			latestToken.UserID = workunit.originalUser.Name
+			_, err = tokenInterface.Update(latestToken)
+			if err != nil {
+				return fmt.Errorf("[%v] unable to update token: %w", migrateTokensOperation, err)
+			}
+		}
+	}
+	return nil
+}
+
+func collectTokens(workunits *[]migrateUserWorkUnit, sc *config.ScaledContext) error {
+	tokenInterface := sc.Management.Tokens("")
+	tokenList, err := tokenInterface.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] unable to fetch token objects: %v", migrateAdUserOperation, err)
+		return err
+	}
+
+	for i, workunit := range *workunits {
+		guidPrincipal := activeDirectoryPrefix + workunit.guid
+		for _, token := range tokenList.Items {
+			if guidPrincipal == token.UserPrincipal.Name || workunit.originalUser.Name == token.UserID {
+				workunit.guidTokens = append(workunit.guidTokens, token)
+			} else {
+				for _, duplicateLocalUser := range workunit.duplicateUsers {
+					if localPrincipalID(duplicateLocalUser) == token.UserPrincipal.Name {
+						workunit.duplicateLocalTokens = append(workunit.duplicateLocalTokens, token)
+					}
+				}
+			}
+		}
+		(*workunits)[i] = workunit
+	}
+
+	return nil
+}
+
+func workUnitContainsName(workunit *migrateUserWorkUnit, name string) bool {
+	if workunit.originalUser.Name == name {
+		return true
+	}
+	for _, duplicateLocalUser := range workunit.duplicateUsers {
+		if duplicateLocalUser.Name == name {
+			return true
+		}
+	}
+	return false
+}
+
+func collectCRTBs(workunits *[]migrateUserWorkUnit, sc *config.ScaledContext) error {
+	crtbInterface := sc.Management.ClusterRoleTemplateBindings("")
+	crtbList, err := crtbInterface.List(metav1.ListOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] unable to fetch CRTB objects: %v", migrateAdUserOperation, err)
+		return err
+	}
+
+	// first build a map of guid-principalid -> work unit, which will make the following logic more efficient
+	originalGUIDWorkUnits := map[string]int{}
+	duplicateGUIDWorkUnits := map[string]int{}
+	for i, workunit := range *workunits {
+		originalGUIDWorkUnits[activeDirectoryPrefix+workunit.guid] = i
+		for j := range workunit.duplicateUsers {
+			duplicateGUIDWorkUnits[activeDirectoryPrefix+workunit.guid] = j
+		}
+	}
+
+	for _, crtb := range crtbList.Items {
```

Comment:
I'm not sure that this loop (and the corresponding one for the PRTBs) will work for local users. In those cases, the `crtb.UserPrincipalName` could be `local://user-123xyz` - your map is GUID only, so you aren't going to see those bindings.